nixos/resolvconf: allow different implementations

ncfavier opened the pull request:
Fixes https://github.com/NixOS/nixpkgs/issues/139526 and possibly https://github.com/NixOS/nixpkgs/pull/118742#issuecomment-816929646 by adding a `networking.resolvconf.package` option that can be either openresolv or systemd depending on the chosen implementation.

Weakly depends on
- https://github.com/NixOS/nixpkgs/pull/163958 and
- https://github.com/NixOS/nixpkgs/pull/166142

for the other wg-quick improvements.

Note: networkd users should probably be encouraged to define their wireguard config in networkd directly, but there's no reason the wg-quick module shouldn't also work for now.

ncfavier committed:
nixos/resolvconf: add `package`

Expose the package that provides the system-wide `resolvconf` command
(either openresolv or systemd) to allow implementation-agnostic modules.

ncfavier committed:
wireguard-tools: use resolvconf from PATH if available

Allows using wg-quick with systemd-resolved which provides its own
implementation of the resolvconf command.

ncfavier committed:
nixos/wg-quick: improve usage with systemd-networkd

Use `networking.resolvconf.package` to allow DNS entries to be set using
the system-wide resolver implementation instead of hardcoding systemd or
openresolv.

Extend the tests by adding DNS entries and making one of the peers use
systemd-networkd (hence systemd-resolved).

Also add a few `networkd`-specific settings.

ncfavier committed:
nixos/tailscale: use `networking.resolvconf.package`

Allow using the appropriate resolver implementation to set DNS entries
(typically systemd-resolved instead of openresolv).

SuperSandro2000 commented on nixos/modules/config/resolvconf.nix:
```diff
@@ -54,6 +54,14 @@ in
 
+      package = mkOption {
+        type = types.package;
+        internal = true;
```
Not sure why this is internal.

ncfavier replied:
No good reason.

ncfavier replied:
Well, I think my intention was that users shouldn't see this in the manual and think they have to set it. Maybe just say that in the description?

SuperSandro2000 replied:
I don't know. A good description and default value can get you a long way.

ncfavier replied:
I extended the description and set `openresolv` as the default.

ncfavier commented:
Dependencies resolved!

ncfavier commented:
@ofborg test wireguard

ncfavier committed:
nixos/dhcpcd: use `networking.resolvconf.package`

Allow using the appropriate resolver implementation to set DNS entries
(typically systemd-resolved instead of openresolv).

ncfavier commented:
I've added a commit to switch dhcpcd to `networking.resolvconf.package`, could you test this please?

@ofborg test wireguard networking

SuperSandro2000 commented:
Just tested this locally and this fixes wg-quick when used together with systemd-resolved.

I am going ahead and merging this now.

ncfavier commented:
Thanks!
